designs/0035-pre-model-gqs.md

```diff
@@ -0,0 +1,678 @@
+methods for I/O that depend on features of the model.  It is defined
+to extend the `model_base` class, which is used by the external
+interfaces.
+
+To deal with construction, the generated C++ code for a model also
+includes a generic factory method to return a base model.
+
+```cpp
+stan::model::model_base&
+new_model(stan::io::var_context& data_context, unsigned int seed,
+          std::ostream* msg_stream);
+		  ```
+
+This simply creates an instance of `foo_model` with the specified
+arguments and returns it.
+
+For reference, here are links to the full definitions from the
+`stan-dev/stan` GitHub repository:
+
+* [`model_base.hpp`](https://github.com/stan-dev/stan/blob/develop/src/stan/model/model_base.hpp)
+* [`model_base_crtp.hpp`](https://github.com/stan-dev/stan/blob/develop/src/stan/model/model_base_crtp.hpp)
+
+### Representing the latent data
+
+The design is based on an object-based representation of the latent
+data using code generation.  For example, suppose the latent data
+block is defined as follows.
+
+
+```cpp
+latent data {
+  real<lower=0, upper=1> sens = beta_rng(98, 2);
+  real<lower=0, upper=1> spec = beta_rng(95, 5);
+}
+```
+
+This will lead to the code generation of a class in the generated
+`.hpp` file for the model following the CRTP.
+
+```cpp
+struct latent_data : public latent_data_base_crtp<latent_data> {
+  int iteration_number__;  // included by default in all latent_data
+  double sens;
+  double spec;
+};
+```
+
+Here, the `latent_data_base_crtp` will extend `latent_data_base`,
+following the pattern used for the model class.
+
+### Constructing the latent data generically
+
+Just like the model class, this will require a generic factory method,
+though in this case, no arguments are required as it will be used like
+a simple `struct`.
+
+```cpp
+latent_data_base& new_latent_data();
+```
+
+This must then get passed where necessary and cast into the relevant
+type in the class definition, as shown in the next section.
+
+
+### Generating the latent data
+
+An additional method in the model class will be required to populate 
+the latent data. 
+
+```cpp
+template <typename RNG> inline void
+set_latent_data(
+    const VectorXd& params_r, 
+    latent_data_base& ldb, 
+    RNG& base_rng, 
+    std::ostream* pstream=nullptr) const {
+  latent_data& latent_data__ = static_cast<latent_data&>(ldb);
+  ... code generated to populate latent_data__ ...
+}
+```
+
+Because it does not use anything specific to the specific model class,
+this can be implemented inthe `latent_data_base_crtp` class using the
+CRTP.
+
+```
+template <class Derived>
+class latent_data_base_crtp<Derived> {
+template <typename RNG> inline void
+set_latent_data(
+    const VectorXd& params_r, 
+    latent_data_base& ldb, 
+    RNG& base_rng, 
+    std::ostream* pstream=nullptr) const {
+	static_cast<Derived&>(ldb).set_latent_data(params_r, ldb,
+                                               base_rng, pstream);
+};
+```
```

Comment:
Since it's just holding data that is made once I think we can use an allocator scheme instead. So we define a latent data struct inside of the model class which will also use the allocator for acquiring memory.
```c++
struct latent_data {
  int iteration_number__;  // included by default in all latent_data
  // Will hold memory owned by allocator
  Eigen::Map<Eigen::Matrix<double, -1, 1>> sens; 
  std::vector<double, stan::math::local_allocator> spec;  
};
```

For the allocator, we will then also make a member function in the model block that will tell us the size of the latent data  struct.

```c++
std::size_t get_latent_size_bytes() {
  // Assume `sens` is size 5 and spec is size 6
  return sizeof(latent_data) + 5 * sizeof(double) + 6 * sizeof(double);
}
```

The allocator will look very similar to Stan's arena allocator, but instead of storing a vector of unsigned char pointers this will only store one void pointer. We will use just one void pointer so that it's easy to cast the memory inside of the allocator to the `latent_data` class when we need it.
```c++
struct local_allocator {
  void* mem_;
  // Must be fixed size
  local_allocator(std::size_t init_bytes) : mem_(malloc(init_bytes));
  template<typename T>
  void* allocate(std::size_t amount);

  template <typename T>
  T* cast_this() {
    return reinterpret_cast<T*>(mem_);
  }
};
```

Since the sizes of variables have to come from data and transformed data we can know at initialization the amount of memory we need. 

```c++
stan::math::local_allocator latent_alloc(model->get_latent_size_bytes());
```

Then the model class will have a `latent_data` member function that uses the allocator as an in/out parameter

```c++
inline void
set_latent_data(stan::math::allocator& alloc) {
  latent_data* = new (alloc.allocate<latent_data>()) latent_data{0, 
    Eigen::Map<Eigen::Matrix<double, -1, 1>>(new (alloc.allocate<double>(5), 5, 1))
    std::vector<double, stan::math::allocator>(5)
  };
  // Fill in latent_data...
  return;
}
```

So now we can have a new `log_prob` that can take in the allocator. The log_prob will call the allocators `cast` function to get back the pointer to the `latent_data`. We can either use the `latent_data` pointer for the rest of the code or pull out references to the members of `latent_data`.

```c++
auto log_prob(stan::math::allocator& alloc, ...) {
  latent_data* = alloc.cast<latent_data>();
  // Just makes a reference, very low cost
  auto&& spec = latent_data->spec;
  // Rest of the program written as normal
  auto X = Y * spec;
}
```

Reply:
> Since it's just holding data that is made once I think we can use an allocator scheme instead.

What's the advantage to doing it the way you propose?  It looks very similar to me, so I'm sure I'm missing something important.

Reply:
It removes inheritance, does not need a generic factory method, and goes more into how to handle dynamically allocated objects. I don't think we need to have a base class that we upcast to the actual class. Since this is just used in the model class the rest of the stan code can just worry about passing around an allocator that holds the memory for the latent data struct.